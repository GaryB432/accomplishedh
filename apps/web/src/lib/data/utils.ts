import { existsSync, readFileSync } from "node:fs";

export interface FormDataHuman {
  adultbasic: string;
  adultregion: string;
  death: string;
  dob: string;
  era: string;
  id: string;
  inventory: string;
  knownFor: string;
  name: string;
  osfName: string;
  serial: string;
  yob: string;
}

export function getShardPathFor(
  k: string,
  patern = "../../data/shards/*.json",
): string {
  const m = patern.split("*");
  const n = m.join(k ?? "X");
  return n;
}

export function readJSONSync<T>(fn: string, defaultValue?: T): T | undefined {
  // console.log(existsSync.toString());
  if (existsSync(fn)) {
    return JSON.parse(readFileSync(fn, "utf-8"));
  }
  console.log(`no "${fn}" using default`, defaultValue);
  return defaultValue;
}

export function readArraySync<T>(fn: string, defaultValue?: T[]): Array<T> {
  if (!existsSync(fn)) {
    if (defaultValue) {
      return defaultValue;
    }
    throw new Error(`${fn} seems to not exist`);
  }
  return JSON.parse(readFileSync(fn, "utf-8"));
}
