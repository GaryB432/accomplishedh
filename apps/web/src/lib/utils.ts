import crypto from "crypto-js";

export function isGary(locals: App.Locals): boolean {
  return (
    crypto.MD5(locals.userid).toString(crypto.enc.Hex) ===
    "03b083fd0aadc8883198881ba88111ab"
  );
}

export async function delay<T>(cb: () => Promise<T>, seconds = 1): Promise<T> {
  return new Promise((success) => {
    setTimeout(() => {
      success(cb());
    }, seconds * 1000);
  });
}
