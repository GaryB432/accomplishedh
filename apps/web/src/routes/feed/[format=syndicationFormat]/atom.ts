import {
  encodeHTML,
  isValidISO8601,
  midday,
  type FeaturedHuman,
} from "@accomplishedh/shared";
import * as path from "node:path";
import { type XmlObject } from "xml";

function encodedHTML(s: string): [XmlObject, string] {
  return [{ _attr: { type: "html" } }, encodeHTML(s)];
}

export async function feed(
  hs: Readonly<FeaturedHuman>[],
  baseHref: string,
  todayIso: string,
): Promise<XmlObject> {
  if (!isValidISO8601(todayIso)) {
    throw new Error(`"${todayIso}" is not a date we can use atm`);
  }

  const today = midday(todayIso);

  const featuredEntries = hs
    .toReversed()
    .filter((h) => !(h.stamp > today))
    .map((h) => {
      const updated = midday(h.stamp);
      const entry = {
        entry: [
          { id: `urn:uri:${h.id}` },
          { title: `${h.human.name}` },
          { updated },
          {
            link: [
              {
                _attr: {
                  rel: "alternate",
                  type: "text/html",
                  href: new URL(
                    path.join("featured", h.id.slice(0, 8)),
                    baseHref,
                  ).href,
                },
              },
            ],
          },
          {
            link: [
              {
                _attr: {
                  rel: "enclosure",
                  type: "image/jpeg",
                  href: "https:".concat(h.human.portrait.img["src"]),
                },
              },
            ],
          },
          {
            content: encodedHTML(
              `<h1>${h.human.name}</h1><div>${h.human.knownFor}</div>`,
            ),
          },
        ],
      };
      return entry;
    });

  const feed = [
    { _attr: { xmlns: "http://www.w3.org/2005/Atom" } },
    { title: [{ _attr: { type: "text" } }, "Human Accomplishment"] },
    {
      subtitle: encodedHTML(
        "Bringing you the <em>most</em> accomplished humans every day",
      ),
    },

    { updated: today },

    {
      id: baseHref,
    },
    {
      link: [
        {
          _attr: {
            rel: "alternate",
            type: "text/html",
            hreflang: "en",
            href: baseHref,
          },
        },
      ],
    },
    {
      link: [
        {
          _attr: {
            rel: "self",
            type: "application/atom+xml",
            href: new URL("feed/atom", baseHref).href,
          },
        },
      ],
    },
    { author: { name: "Human Accomplishment Editors" } },
    {
      author: [
        { name: "Human Accomplishment Editors" },
        { email: "editor@accomplishedh.com" },
      ],
    },
    {
      rights: `Copyright (c) ${todayIso.slice(0, 4)}, Human Accomplishment`,
    },
    ...featuredEntries,
  ];
  return { feed };
}
