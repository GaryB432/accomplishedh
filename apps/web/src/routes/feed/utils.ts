export function homeOf(href: string): string {
  const { origin, host, hostname, pathname } = URL.parse(href)!;
  if (
    hostname === "localhost" ||
    hostname.endsWith("humanaccomplishment.com") ||
    hostname.endsWith(".app")
  ) {
    return origin;
  }

  throw new Error(
    `${JSON.stringify({ origin, host, hostname, pathname })} Unrecognized above`,
  );
}
