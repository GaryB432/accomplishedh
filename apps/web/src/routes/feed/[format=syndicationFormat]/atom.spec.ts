import type { EuroHuman, FeaturedHuman } from "@accomplishedh/shared";
import { describe, expect, test } from "vitest";

import { feed } from "./atom";
import { homeOf } from "../utils";

// vi.mock('xml', () => {
//   return {
//     default: vi.fn(() => {
//       return `<sorry>nope</sorry>`;
//     }),
//   };
// });

describe("Atom", () => {
  const JUL95 = "2095-07-31";

  const yesterday = new Date(Date.parse(JUL95) - 1 * 24 * 60 * 60 * 1000);

  const someNearbyDays = Array(5)
    .fill(0)
    .map((_, n) => {
      const nd = new Date(yesterday);
      nd.setDate(nd.getDate() + n);
      return nd.toISOString().slice(0, 10);
    });

  test("feed", async () => {
    const atom = await feed(
      someNearbyDays.flatMap<FeaturedHuman>((stamp, dn) => {
        return Array(3)
          .fill(0)
          .map((_, n) => {
            const seq = dn * 3 + n;
            const dseq = seq.toString().padStart(6, "0");
            const name = `SUBJECT ${seq.toString().padStart(6, "0")}`;
            return {
              human: {
                ...defh,
                serial: dseq,
                id: dseq.padStart(10, "0"),
                name,
                knownFor: `a guy featured on ${stamp}`,
                wikibase: { entity: { id: `Q${dseq}` } },
                portrait: {
                  above: "Rudolf Albert von Kölliker",
                  img: {
                    alt: "Kölliker Rudolph Albert von 1818-1902.jpg",
                    src: `//upload.wikimedia.org/wikipedia/commons/thumb/p${seq}.jpg`,
                    width: "250",
                    height: "335",
                  },
                },
              },
              id: `dfh${seq}`,
              stamp,
            };
          });
      }),
      homeOf("https://projects.hoster.app/feed/atom"),
      JUL95,
    );
    expect(atom).toMatchSnapshot();
  });
});

const defh: EuroHuman = {
  adultbasic: "Britain",
  adultregion: "E.SE",
  yob: "1903",
  death: "1975-05-20T00:00:00.000Z",
  dob: "01-10",
  era: "1800+",
  id: "",
  inventory: "Art.West",
  knownFor: "",
  name: "",
  osfName: "",
  portrait: {
    img: { fun: "yep" },
  },
  props: [],
  serial: "-1",
  sr: undefined,
};

// <?xml version="1.0" encoding="utf-8"?>
//    <feed xmlns="http://www.w3.org/2005/Atom">
//      <title type="text">dive into mark</title>
//      <subtitle type="html">
//        A &lt;em&gt;lot&lt;/em&gt; of effort
//        went into making this effortless
//      </subtitle>
//      <updated>2005-07-31T12:29:29Z</updated>
//      <id>tag:example.org,2003:3</id>
//      <link rel="alternate" type="text/html"
//       hreflang="en" href="http://example.org/"/>
//      <link rel="self" type="application/atom+xml"
//       href="http://example.org/feed.atom"/>
//      <rights>Copyright (c) 2003, Mark Pilgrim</rights>
//      <generator uri="http://www.example.com/" version="1.0">ddd
//        Example Toolkit
//      </generator>
//      <entry>
//        <title>Atom draft-07 snapshot</title>
//        <link rel="alternate" type="text/html"
//         href="http://example.org/2005/04/02/atom"/>
//        <link rel="enclosure" type="audio/mpeg" length="1337"
//         href="http://example.org/audio/ph34r_my_podcast.mp3"/>
//        <id>tag:example.org,2003:3.2397</id>
//        <updated>2005-07-31T12:29:29Z</updated>
//        <published>2003-12-13T08:29:29-04:00</published>
//        <author>
//          <name>Mark Pilgrim</name>
//          <uri>http://example.org/</uri>
//          <email>f8dy@example.com</email>
//        </author>
//        <contributor>
//          <name>Sam Ruby</name>
//        </contributor>
//        <contributor>
//          <name>Joe Gregorio</name>
//        </contributor>
//        <content type="xhtml" xml:lang="en"
//         xml:base="http://diveintomark.org/">
//          <div xmlns="http://www.w3.org/1999/xhtml">
//            <p><i>[Update: The Atom draft is finished.]</i></p>
//          </div>
//        </content>
//      </entry>
//    </feed>
