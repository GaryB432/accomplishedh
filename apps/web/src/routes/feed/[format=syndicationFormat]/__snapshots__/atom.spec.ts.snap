// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Atom > feed 1`] = `
{
  "feed": [
    {
      "_attr": {
        "xmlns": "http://www.w3.org/2005/Atom",
      },
    },
    {
      "title": [
        {
          "_attr": {
            "type": "text",
          },
        },
        "Human Accomplishment",
      ],
    },
    {
      "subtitle": [
        {
          "_attr": {
            "type": "html",
          },
        },
        "Bringing you the &lt;em&gt;most&lt;/em&gt; accomplished humans every day",
      ],
    },
    {
      "updated": "2095-07-31T12:00:00.000Z",
    },
    {
      "id": "https://projects.hoster.app",
    },
    {
      "link": [
        {
          "_attr": {
            "href": "https://projects.hoster.app",
            "hreflang": "en",
            "rel": "alternate",
            "type": "text/html",
          },
        },
      ],
    },
    {
      "link": [
        {
          "_attr": {
            "href": "https://projects.hoster.app/feed/atom",
            "rel": "self",
            "type": "application/atom+xml",
          },
        },
      ],
    },
    {
      "author": {
        "name": "Human Accomplishment Editors",
      },
    },
    {
      "author": [
        {
          "name": "Human Accomplishment Editors",
        },
        {
          "email": "editor@accomplishedh.com",
        },
      ],
    },
    {
      "rights": "Copyright (c) 2095, Human Accomplishment",
    },
    {
      "entry": [
        {
          "id": "urn:uri:dfh5",
        },
        {
          "title": "SUBJECT 000005",
        },
        {
          "updated": "2095-07-31T12:00:00.000Z",
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://projects.hoster.app/featured/dfh5",
                "rel": "alternate",
                "type": "text/html",
              },
            },
          ],
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://upload.wikimedia.org/wikipedia/commons/thumb/p5.jpg",
                "rel": "enclosure",
                "type": "image/jpeg",
              },
            },
          ],
        },
        {
          "content": [
            {
              "_attr": {
                "type": "html",
              },
            },
            "&lt;h1&gt;SUBJECT 000005&lt;/h1&gt;&lt;div&gt;a guy featured on 2095-07-31&lt;/div&gt;",
          ],
        },
      ],
    },
    {
      "entry": [
        {
          "id": "urn:uri:dfh4",
        },
        {
          "title": "SUBJECT 000004",
        },
        {
          "updated": "2095-07-31T12:00:00.000Z",
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://projects.hoster.app/featured/dfh4",
                "rel": "alternate",
                "type": "text/html",
              },
            },
          ],
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://upload.wikimedia.org/wikipedia/commons/thumb/p4.jpg",
                "rel": "enclosure",
                "type": "image/jpeg",
              },
            },
          ],
        },
        {
          "content": [
            {
              "_attr": {
                "type": "html",
              },
            },
            "&lt;h1&gt;SUBJECT 000004&lt;/h1&gt;&lt;div&gt;a guy featured on 2095-07-31&lt;/div&gt;",
          ],
        },
      ],
    },
    {
      "entry": [
        {
          "id": "urn:uri:dfh3",
        },
        {
          "title": "SUBJECT 000003",
        },
        {
          "updated": "2095-07-31T12:00:00.000Z",
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://projects.hoster.app/featured/dfh3",
                "rel": "alternate",
                "type": "text/html",
              },
            },
          ],
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://upload.wikimedia.org/wikipedia/commons/thumb/p3.jpg",
                "rel": "enclosure",
                "type": "image/jpeg",
              },
            },
          ],
        },
        {
          "content": [
            {
              "_attr": {
                "type": "html",
              },
            },
            "&lt;h1&gt;SUBJECT 000003&lt;/h1&gt;&lt;div&gt;a guy featured on 2095-07-31&lt;/div&gt;",
          ],
        },
      ],
    },
    {
      "entry": [
        {
          "id": "urn:uri:dfh2",
        },
        {
          "title": "SUBJECT 000002",
        },
        {
          "updated": "2095-07-30T12:00:00.000Z",
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://projects.hoster.app/featured/dfh2",
                "rel": "alternate",
                "type": "text/html",
              },
            },
          ],
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://upload.wikimedia.org/wikipedia/commons/thumb/p2.jpg",
                "rel": "enclosure",
                "type": "image/jpeg",
              },
            },
          ],
        },
        {
          "content": [
            {
              "_attr": {
                "type": "html",
              },
            },
            "&lt;h1&gt;SUBJECT 000002&lt;/h1&gt;&lt;div&gt;a guy featured on 2095-07-30&lt;/div&gt;",
          ],
        },
      ],
    },
    {
      "entry": [
        {
          "id": "urn:uri:dfh1",
        },
        {
          "title": "SUBJECT 000001",
        },
        {
          "updated": "2095-07-30T12:00:00.000Z",
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://projects.hoster.app/featured/dfh1",
                "rel": "alternate",
                "type": "text/html",
              },
            },
          ],
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://upload.wikimedia.org/wikipedia/commons/thumb/p1.jpg",
                "rel": "enclosure",
                "type": "image/jpeg",
              },
            },
          ],
        },
        {
          "content": [
            {
              "_attr": {
                "type": "html",
              },
            },
            "&lt;h1&gt;SUBJECT 000001&lt;/h1&gt;&lt;div&gt;a guy featured on 2095-07-30&lt;/div&gt;",
          ],
        },
      ],
    },
    {
      "entry": [
        {
          "id": "urn:uri:dfh0",
        },
        {
          "title": "SUBJECT 000000",
        },
        {
          "updated": "2095-07-30T12:00:00.000Z",
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://projects.hoster.app/featured/dfh0",
                "rel": "alternate",
                "type": "text/html",
              },
            },
          ],
        },
        {
          "link": [
            {
              "_attr": {
                "href": "https://upload.wikimedia.org/wikipedia/commons/thumb/p0.jpg",
                "rel": "enclosure",
                "type": "image/jpeg",
              },
            },
          ],
        },
        {
          "content": [
            {
              "_attr": {
                "type": "html",
              },
            },
            "&lt;h1&gt;SUBJECT 000000&lt;/h1&gt;&lt;div&gt;a guy featured on 2095-07-30&lt;/div&gt;",
          ],
        },
      ],
    },
  ],
}
`;
